crates/vm/src/arch/integration_api.rs

```diff
@@ -216,25 +216,29 @@ pub struct AdapterAirContext<T, I: VmAdapterInterface<T>> {
     pub instruction: I::ProcessedInstruction,
 }
 
-/// Interface for trace generation when the state transition step of a single instruction
-/// uses only one trace row. The trace row is provided as a mutable buffer during both
-/// instruction execution and trace generation.
+/// Interface for trace generation of a single instruction.The trace is provided as a mutable
+/// buffer during both instruction execution and trace generation.
 /// It is expected that no additional memory allocation is necessary and the trace buffer
 /// is sufficient, with possible overwriting.
-pub trait SingleTraceStep<F, CTX> {
+pub trait TraceStep<F, CTX> {
     fn execute(
         &mut self,
         state: VmStateMut<TracingMemory, CTX>,
         instruction: &Instruction<F>,
-        row_slice: &mut [F],
+        // TODO(ayush): combine to a single struct
+        trace: &mut [F],
+        trace_offset: &mut usize,
+        // TODO(ayush): move air inside step and remove width
+        width: usize,
     ) -> Result<()>;
 
     /// Populates `row_slice`. This function will always be called after
-    /// [`SingleTraceStep::execute`], so the `row_slice` should already contain context necessary to
+    /// [`TraceStep::execute`], so the `row_slice` should already contain context necessary to
     /// fill in the rest of the row. This function will be called for each row in the trace which is
     /// being used, and all other rows in the trace will be filled with zeroes.
     ///
     /// The provided `row_slice` will have length equal to the width of the AIR.
+    // TODO(ayush): should this still operate on a single row or be like execute and operate on a set of rows?
```

Comment:
Yea it doesn't make sense if `execute` is on whole trace; rename it to `fill_trace` and provide the whole trace, parallel to `execute`?

Reply:
done